File: frontend/src/utils/sanitization.ts
/**
 * Client-side input sanitization for JT-Note.
 * Provides real-time sanitization for all user inputs before they are sent to the API.
 * Defense in depth: client sanitizes for UX feedback, server re-validates.
 */

// Maximum lengths matching server-side constants
const MAX_DISPLAY_NAME = 100;
const MAX_GROUP_NAME = 100;
const MAX_STATUS_TEXT = 200;
const MAX_MESSAGE_CONTENT = 10000;
const MAX_USERNAME = 30;

/**
 * Strip control characters from text.
 * Prevents terminal injection and BCP47 spoofing attacks.
 */
function stripControlChars(text: string): string {
  // Remove characters \x00-\x08, \x0B, \x0C, \x0E-\x1F, \x7F
  return text.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g, '');
}

/**
 * HTML-encode special characters to prevent XSS when text is rendered.
 * Converts: < > & " ' to their HTML entity equivalents.
 */
function htmlEncode(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;');
}

/**
 * Sanitize display text (names, group names, status).
 * Applied before sending to API and before rendering.
 */
export function sanitizeDisplayText(text: string, maxLength: number = MAX_DISPLAY_NAME): string {
  if (!text) return '';
  let cleaned = text.trim();
  cleaned = stripControlChars(cleaned);
  cleaned = cleaned.slice(0, maxLength);
  return cleaned;
}

/**
 * Sanitize for safe HTML rendering — returns HTML-encoded string.
 * Use this when displaying user-generated content in any HTML/text context.
 */
export function sanitizeForRender(text: string): string {
  if (!text) return '';
  let cleaned = stripControlChars(text);
  return htmlEncode(cleaned);
}

/**
 * Validate and sanitize username input.
 * Strict allowlist: alphanumeric, underscore, hyphen, dot.
 */
export function sanitizeUsername(text: string): string {
  let cleaned = text.trim().toLowerCase();
  cleaned = cleaned.replace(/[^a-z0-9_\-.]/g, '');
  return cleaned.slice(0, MAX_USERNAME);
}

/**
 * Sanitize group name.
 */
export function sanitizeGroupName(text: string): string {
  return sanitizeDisplayText(text, MAX_GROUP_NAME);
}

/**
 * Sanitize status text.
 */
export function sanitizeStatusText(text: string): string {
  return sanitizeDisplayText(text, MAX_STATUS_TEXT);
}

/**
 * Sanitize message content before sending.
 * Note: E2EE messages are encrypted, but this applies to plaintext metadata.
 */
export function sanitizeMessageContent(text: string): string {
  return sanitizeDisplayText(text, MAX_MESSAGE_CONTENT);
}

/**
 * Sanitize search query — strip regex special chars.
 */
export function sanitizeSearchQuery(text: string): string {
  let cleaned = text.trim();
  cleaned = stripControlChars(cleaned);
  // Escape regex special characters
  cleaned = cleaned.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  return cleaned.slice(0, 200);
}

/**
 * Validate callsign: alphanumeric + hyphen only.
 */
export function sanitizeCallsign(text: string): string {
  let cleaned = text.trim().toUpperCase();
  cleaned = cleaned.replace(/[^A-Z0-9\-]/g, '');
  return cleaned.slice(0, 20);
}

/**
 * Validate that a string is not empty after sanitization.
 */
export function isEmptyAfterSanitization(text: string): boolean {
  return !sanitizeDisplayText(text).trim();
}

/**
 * React Native safe text renderer.
 * For React Native, Text components don't execute HTML, but this
 * ensures consistency and protects against any future web-view rendering.
 */
export function safeText(text: string | null | undefined): string {
  if (!text) return '';
  return stripControlChars(text);
}

File: backend/sanitization.py
"""
Input sanitization utilities for JT-Note backend.
All user-facing text fields pass through these functions before DB storage.
Defense in depth: Pydantic validates schema, sanitization cleans content.
"""

import re
import html

# Allowlist: characters permitted in display names, group names, status texts
# No angle brackets, no quotes, no script-related chars
SAFE_DISPLAY_CHARS = re.compile(
    r"^[\w\s\-_.,!?@#&+=$%(){}\[\]:;\"'/\\*\u00C0-\u024F\u0400-\u04FF\u0600-\u06FF\u4E00-\u9FFF\u3040-\u309F\u30A0-\u30FF\ud800-\udbff\udc00-\udfff]*$"
)

# Strip control characters (except newline and tab)
CONTROL_CHARS = re.compile(r"[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]")

# Max lengths per field type (defense against oversized payloads)
MAX_DISPLAY_NAME = 100
MAX_GROUP_NAME = 100
MAX_STATUS_TEXT = 200
MAX_MESSAGE_CONTENT = 10000
MAX_USERNAME = 30


def strip_control_chars(text: str) -> str:
    """Remove control characters that could be used for terminal injection or BCP47 attacks."""
    return CONTROL_CHARS.sub("", text)


def sanitize_display_text(text: str, max_length: int = MAX_DISPLAY_NAME) -> str:
    """
    Sanitize user-facing display text (names, group names, status).
    - Strips control characters
    - HTML-encodes special characters (output encoding)
    - Trims whitespace
    - Enforces max length
    - Rejects if empty after sanitization
    """
    if not text:
        return ""
    cleaned = text.strip()
    cleaned = strip_control_chars(cleaned)
    # HTML encode: < > & " ' become safe entities
    cleaned = html.escape(cleaned, quote=True)
    # Truncate to max length
    cleaned = cleaned[:max_length]
    return cleaned


def sanitize_message_content(text: str, max_length: int = MAX_MESSAGE_CONTENT) -> str:
    """
    Sanitize message content. Messages are E2EE-encrypted so server can't read them,
    but this applies to plaintext fallback and metadata.
    Same sanitization as display text but with higher max length.
    """
    return sanitize_display_text(text, max_length)


def validate_and_sanitize_username(username: str) -> str:
    """
    Username validation: strict allowlist.
    Usernames are already validated via regex in Pydantic, but we add sanitization as defense in depth.
    """
    cleaned = username.strip().lower()
    cleaned = strip_control_chars(cleaned)
    if len(cleaned) > MAX_USERNAME:
        cleaned = cleaned[:MAX_USERNAME]
    return cleaned


def sanitize_poll_text(text: str, max_length: int = 500) -> str:
    """Sanitize poll questions and options."""
    return sanitize_display_text(text, max_length)


def sanitize_search_query(query: str, max_length: int = 200) -> str:
    """
    Sanitize search queries to prevent NoSQL injection via regex.
    Strips regex special characters that could be used in $regex queries.
    """
    cleaned = query.strip()
    cleaned = strip_control_chars(cleaned)
    # Escape regex special characters to prevent NoSQL regex injection
    cleaned = re.escape(cleaned)
    cleaned = cleaned[:max_length]
    return cleaned
